In [ ]:
# Job Market Intelligence System: Problem Statement

## 1. Context & Problem

The current job market is fragmented and opaque, creating significant inefficiencies for three key stakeholder groups:

- **Job Seekers** face information overload, skill uncertainty, and lack of salary transparency.
- **HR Professionals & Recruiters** struggle with competitive hiring, compensation benchmarking, and identifying skill gaps.
- **Educational Institutions & Career Counselors** operate with outdated curriculum and lack real-time market data for guidance.

**Core Problem:** There is no unified, data-driven system that transforms raw job posting data into actionable, real-time insights for all stakeholders.

## 2. Project Goal

To develop a **Job Market Intelligence System** that analyzes job posting data to generate clear, actionable insights on skill demand, geographic opportunity, salary benchmarks, and market trends.

## 3. Key Objectives

1.  **Skill Demand Analysis:** Identify trending and declining technical skills.
2.  **Geographic Opportunity Mapping:** Visualize job distribution and hotspots.
3.  **Salary Benchmarking:** Estimate compensation by role, experience, and location.
4.  **Job Classification & Trend Identification:** Categorize postings and spot emerging roles.

## 4. Primary Business Questions

- **For Job Seekers:** "What skills should I learn, where are the jobs, and what salary can I expect?"
- **For HR/Recruiters:** "How competitive is the market, and are our offers aligned?"
- **For Educators:** "Which skills and emerging roles should we teach for?"

## 5. Success Metrics

- **Technical:** >80% classification accuracy; <$15k MAE for salary prediction.
- **Business:** Delivery of actionable insights, clear visualizations, and identifiable market patterns to all stakeholder groups.

## 6. Project Scope

**In-Scope (Initial Focus):**
- Analysis of provided job posting datasets.
- Focus on English-language technical/professional roles.
- Skills extraction and trend analysis from job descriptions.

**Value Delivered:**
- **Job Seekers:** Reduced search time, clearer career paths.
- **HR Professionals:** Competitive intelligence, optimized recruitment.
- **Educators:** Data-driven curriculum alignment and career guidance.

# Data Exploration and Quality Assessment

Let us now explore our dataset and understand its structure, quality and potential for our project.

## Data Loading and Inspection
We will now load the data andexamine its basic properties.

In [1]:
import pandas as pd
Job_Posting_df = pd.read_csv("Job_Posting_data.csv")
Job_Posting_df.head()

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xd0 in position 1736: invalid continuation byte

- We got an error when trying to read in the dataset due to the unique encoding of the data inside the dataset. Therefore, we had to employ some encoding to debug the dataset and make it readable by the pandas library.

In [2]:
import pandas as pd

encodings_to_try = ['ISO-8859-1', 'cp1252', 'latin1', 'windows-1252', 'utf-8-sig', 'mac_roman']

print("Trying different encodings...")
for encoding in encodings_to_try:
    try:
        Job_Posting_df = pd.read_csv("Job_Posting_data.csv", encoding=encoding)
        print(f"✅ SUCCESS with {encoding} encoding!")
        print(f"   Shape: {Job_Posting_df.shape}")
        print(f"   Columns: {len(Job_Posting_df.columns)}")
        print(f"\nFirst 3 rows:")
        print(Job_Posting_df.head(3))
        print("\nColumn names:")
        for i, col in enumerate(Job_Posting_df.columns, 1):
            print(f"  {i:2}. {col}")
        break
    except UnicodeDecodeError as e:
        print(f"❌ Failed with {encoding}: {str(e)[:50]}...")
    except Exception as e:
        print(f"❌ Failed with {encoding}: {type(e).__name__}")

Trying different encodings...
✅ SUCCESS with ISO-8859-1 encoding!
   Shape: (9919, 21)
   Columns: 21

First 3 rows:
  Website Domain  Ticker                                  Job Opening Title  \
0      bosch.com     NaN  IN_RBAI_Assistant Manager_Dispensing Process E...   
1      bosch.com     NaN  Professional Internship: Hardware Development ...   
2         zf.com     NaN                      Process Expert BMS Production   

                                     Job Opening URL         First Seen At  \
0  https://jobs.smartrecruiters.com/BoschGroup/74...  2024-05-29T19:59:45Z   
1  https://jobs.smartrecruiters.com/BoschGroup/74...  2024-05-04T01:00:12Z   
2  https://jobs.zf.com/job/Shenyang-Process-Exper...  2024-04-19T06:47:24Z   

           Last Seen At                 Location  \
0  2024-07-31T14:35:44Z   Indiana, United States   
1  2024-07-29T17:46:16Z  Delaware, United States   
2  2024-05-16T02:25:08Z                    China   

                                       Locat

In [3]:
Job_Posting_df.head()

Website Domain  Ticker                                  Job Opening Title  \
0      bosch.com     NaN  IN_RBAI_Assistant Manager_Dispensing Process E...   
1      bosch.com     NaN  Professional Internship: Hardware Development ...   
2         zf.com     NaN                      Process Expert BMS Production   
3      bosch.com     NaN  DevOps Developer with Python for ADAS Computin...   
4      bosch.com     NaN  Senior Engineer Sales - Video Systems and Solu...   

                                     Job Opening URL         First Seen At  \
0  https://jobs.smartrecruiters.com/BoschGroup/74...  2024-05-29T19:59:45Z   
1  https://jobs.smartrecruiters.com/BoschGroup/74...  2024-05-04T01:00:12Z   
2  https://jobs.zf.com/job/Shenyang-Process-Exper...  2024-04-19T06:47:24Z   
3  https://jobs.smartrecruiters.com/BoschGroup/74...  2024-08-16T10:20:37Z   
4  https://jobs.smartrecruiters.com/BoschGroup/74...  2024-07-01T17:31:20Z   

           Last Seen At                 Location  \
0  2024-07-31T14:35:44Z   Indiana, United States   
1  2024-07-29T17:46:16Z  Delaware, United States   
2  2024-05-16T02:25:08Z                    China   
3  2024-08-22T11:14:49Z                  Romania   
4  2024-08-01T05:11:33Z                    India   

                                       Location Data  \
0  [{"city":null,"state":"Indiana","zip_code":nul...   
1  [{"city":null,"state":"Delaware","zip_code":nu...   
2  [{"city":null,"state":null,"zip_code":null,"co...   
3  [{"city":null,"state":null,"zip_code":null,"co...   
4  [{"city":null,"state":null,"zip_code":null,"co...   

                                       Category    Seniority  ...  \
0              engineering, management, support      manager  ...   
1                                    internship  non_manager  ...   
2                                   engineering  non_manager  ...   
3  information_technology, software_development  non_manager  ...   
4                            engineering, sales  non_manager  ...   

                                         Description Salary  \
0  **IN\_RBAI\_Assistant Manager\_Dispensing Proc...    NaN   
1  **Professional Internship: Hardware Developmen...    NaN   
2  ZF is a global technology company supplying sy...    NaN   
3  **DevOps Developer with Python for ADAS Comput...    NaN   
4  **Senior Engineer Sales - Video Systems and So...    NaN   

                                         Salary Data  \
0  {"salary_low":null,"salary_high":null,"salary_...   
1  {"salary_low":null,"salary_high":null,"salary_...   
2  {"salary_low":null,"salary_high":null,"salary_...   
3  {"salary_low":null,"salary_high":null,"salary_...   
4  {"salary_low":null,"salary_high":null,"salary_...   

               Contract Types Job Status Job Language Job Last Processed At  \
0                   full time     closed           en  2024-08-02T14:47:55Z   
1  full time, internship, m/f     closed           en  2024-07-31T17:50:07Z   
2                         NaN     closed           en  2024-05-18T02:32:04Z   
3                   full time     closed           en  2024-08-23T00:33:30Z   
4                   full time     closed           en  2024-08-02T19:03:16Z   

   O*NET Code                       O*NET Family  \
0  43-1011.00  Office and Administrative Support   
1  17-2061.00       Architecture and Engineering   
2  51-9141.00                         Production   
3  15-1252.00          Computer and Mathematical   
4  41-9031.00                  Sales and Related   

                               O*NET Occupation Name  
0  First-Line Supervisors of Office and Administr...  
1                        Computer Hardware Engineers  
2               Semiconductor Processing Technicians  
3                                Software Developers  
4                                    Sales Engineers  

[5 rows x 21 columns]

In [6]:
Job_Posting_df.shape

(9919, 21)

In [7]:
Job_Posting_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9919 entries, 0 to 9918
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Website Domain         9919 non-null   object 
 1   Ticker                 0 non-null      float64
 2   Job Opening Title      9919 non-null   object 
 3   Job Opening URL        9919 non-null   object 
 4   First Seen At          9919 non-null   object 
 5   Last Seen At           9919 non-null   object 
 6   Location               9508 non-null   object 
 7   Location Data          9919 non-null   object 
 8   Category               8250 non-null   object 
 9   Seniority              9919 non-null   object 
 10  Keywords               7646 non-null   object 
 11  Description            9807 non-null   object 
 12  Salary                 576 non-null    object 
 13  Salary Data            9919 non-null   object 
 14  Contract Types         8004 non-null   object 
 15  Job 

- We observed that there were 21 columns present in the dataset and 9919 rows. We also observed that one column, **Ticker** was a null column which we later dropped while doing the data preparaton.
- We then proceeded to doing EDA.

# Exploratory Data Analysis

- We started by doing an overview of the dataset.

In [8]:
import numpy as np
import json
from datetime import datetime
import warnings
warnings.filterwarnings('ignore')

# Display all columns
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth', 50)

print("="*70)
print("DATASET OVERVIEW")
print("="*70)
print(f"📊 Total Records: {Job_Posting_df.shape[0]:,}")
print(f"📊 Total Features: {Job_Posting_df.shape[1]}")
print(f"📅 Data loaded: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")

DATASET OVERVIEW
📊 Total Records: 9,919
📊 Total Features: 21
📅 Data loaded: 2026-02-12 14:03:02


Columns Summary

In [9]:
print("\n" + "="*70)
print("COLUMN SUMMARY")
print("="*70)
print("\nIndex | Column Name                    | Non-Null | Dtype")
print("-"*60)

for i, col in enumerate(Job_Posting_df.columns, 1):
    non_null = Job_Posting_df[col].notnull().sum()
    percentage = (non_null / len(Job_Posting_df)) * 100
    dtype = Job_Posting_df[col].dtype
    print(f"{i:5d} | {col:30} | {non_null:7,d} ({percentage:5.1f}%) | {dtype}")


COLUMN SUMMARY

Index | Column Name                    | Non-Null | Dtype
------------------------------------------------------------
    1 | Website Domain                 |   9,919 (100.0%) | object
    2 | Ticker                         |       0 (  0.0%) | float64
    3 | Job Opening Title              |   9,919 (100.0%) | object
    4 | Job Opening URL                |   9,919 (100.0%) | object
    5 | First Seen At                  |   9,919 (100.0%) | object
    6 | Last Seen At                   |   9,919 (100.0%) | object
    7 | Location                       |   9,508 ( 95.9%) | object
    8 | Location Data                  |   9,919 (100.0%) | object
    9 | Category                       |   8,250 ( 83.2%) | object
   10 | Seniority                      |   9,919 (100.0%) | object
   11 | Keywords                       |   7,646 ( 77.1%) | object
   12 | Description                    |   9,807 ( 98.9%) | object
   13 | Salary                         |     576 (  5.8%) |

As you can see, our dataset contains 19 columns, one which contains numerical values and the other which are text columns. We will now proceed on data exploration and quality analysis.

 Data Exploration and Quality Assessment

In [7]:
import numpy as np
import json
from datetime import datetime
import warnings
warnings.filterwarnings('ignore')

# Display all columns
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth', 50)

print("="*70)
print("DATASET OVERVIEW")
print("="*70)
print(f"📊 Total Records: {Job_Posting_df.shape[0]:,}")
print(f"📊 Total Features: {Job_Posting_df.shape[1]}")
print(f"📅 Data loaded: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")

DATASET OVERVIEW
📊 Total Records: 9,919
📊 Total Features: 21
📅 Data loaded: 2026-02-14 14:06:51


- Let's now do a column summary;

In [8]:
print("\n" + "="*70)
print("COLUMN SUMMARY")
print("="*70)
print("\nIndex | Column Name                    | Non-Null | Dtype")
print("-"*60)

for i, col in enumerate(Job_Posting_df.columns, 1):
    non_null = Job_Posting_df[col].notnull().sum()
    percentage = (non_null / len(Job_Posting_df)) * 100
    dtype = Job_Posting_df[col].dtype
    print(f"{i:5d} | {col:30} | {non_null:7,d} ({percentage:5.1f}%) | {dtype}")


COLUMN SUMMARY

Index | Column Name                    | Non-Null | Dtype
------------------------------------------------------------
    1 | Website Domain                 |   9,919 (100.0%) | object
    2 | Ticker                         |       0 (  0.0%) | float64
    3 | Job Opening Title              |   9,919 (100.0%) | object
    4 | Job Opening URL                |   9,919 (100.0%) | object
    5 | First Seen At                  |   9,919 (100.0%) | object
    6 | Last Seen At                   |   9,919 (100.0%) | object
    7 | Location                       |   9,508 ( 95.9%) | object
    8 | Location Data                  |   9,919 (100.0%) | object
    9 | Category                       |   8,250 ( 83.2%) | object
   10 | Seniority                      |   9,919 (100.0%) | object
   11 | Keywords                       |   7,646 ( 77.1%) | object
   12 | Description                    |   9,807 ( 98.9%) | object
   13 | Salary                         |     576 (  5.8%) |

- Now that we have done a summary of the columns, let's go ahead and have a look at the number of missing values, since as you can see, the summary we have done above shows us the percentage of non-null values in the respective columns.

In [9]:
# Missing Values Analysis

print("\n" + "="*70)
print("MISSING VALUES ANALYSIS - TOP 10 WORST COLUMNS")
print("="*70)

# Calculate missing values
missing_data = []
for col in Job_Posting_df.columns:
    non_null = Job_Posting_df[col].notnull().sum()
    null_count = Job_Posting_df[col].isnull().sum()
    null_pct = (null_count / len(Job_Posting_df)) * 100
    missing_data.append({
        'Column': col,
        'Non-Null': non_null,
        'Null Count': null_count,
        'Null %': null_pct,
        'Dtype': Job_Posting_df[col].dtype
    })

missing_df = pd.DataFrame(missing_data)
missing_df = missing_df.sort_values('Null %', ascending=False)

# Display top 10
print(missing_df.head(10).to_string(index=False))


MISSING VALUES ANALYSIS - TOP 10 WORST COLUMNS
        Column  Non-Null  Null Count     Null %   Dtype
        Ticker         0        9919 100.000000 float64
        Salary       576        9343  94.192963  object
    Job Status      6772        3147  31.726989  object
      Keywords      7646        2273  22.915616  object
Contract Types      8004        1915  19.306382  object
      Category      8250        1669  16.826293  object
      Location      9508         411   4.143563  object
   Description      9807         112   1.129146  object
  O*NET Family      9916           3   0.030245  object
    O*NET Code      9916           3   0.030245  object


In [10]:
print("\n" + "="*70)
print("MISSING DATA CATEGORIZATION")
print("="*70)

# Categorize columns by missing percentage
def categorize_missing(pct):
    if pct == 0:
        return 'Complete (0%)'
    elif pct < 5:
        return 'Good (<5%)'
    elif pct < 20:
        return 'Moderate (5-20%)'
    elif pct < 50:
        return 'High (20-50%)'
    elif pct < 100:
        return 'Very High (50-99%)'
    else:
        return 'Completely Missing (100%)'

missing_df['Category'] = missing_df['Null %'].apply(categorize_missing)
category_counts = missing_df['Category'].value_counts()

for category, count in category_counts.items():
    cols_in_category = missing_df[missing_df['Category'] == category]['Column'].tolist()
    print(f"\n{category:30}: {count:2d} columns")
    if len(cols_in_category) <= 5:
        print(f"   {', '.join(cols_in_category)}")
    else:
        print(f"   {', '.join(cols_in_category[:3])}, ... and {len(cols_in_category)-3} more")


MISSING DATA CATEGORIZATION

Complete (0%)                 :  9 columns
   Seniority, Location Data, Salary Data, ... and 6 more

Good (<5%)                    :  6 columns
   Location, Description, O*NET Family, ... and 3 more

High (20-50%)                 :  2 columns
   Job Status, Keywords

Moderate (5-20%)              :  2 columns
   Contract Types, Category

Completely Missing (100%)     :  1 columns
   Ticker

Very High (50-99%)            :  1 columns
   Salary


- Now that we have an idea of the missing values and their percentages in the dataset, we can now see that the columns, **Ticker** and **Salary**, can be dropped from our dataset. But instead of going with this approach of dropping columns, let's do a critical column analysis to determine which columns are the most important for our analysis.

In [11]:
# 2.3 Critical Column Assessment

print("\n" + "="*70)
print("CRITICAL COLUMNS ASSESSMENT")
print("="*70)

critical_columns = {
    'Job Opening Title': 'Primary identifier - ESSENTIAL',
    'Description': 'Contains skills/requirements - ESSENTIAL',
    'Category': 'Job classification - IMPORTANT',
    'Location': 'Geographic info - IMPORTANT',
    'Seniority': 'Experience level - IMPORTANT',
    'Salary': 'Compensation - DESIRABLE but limited',
    'Contract Types': 'Job type - DESIRABLE',
    'Job Status': 'Open/Closed status - DESIRABLE'
}

print("\nColumn                  | Non-Null |   %   | Status")
print("-"*60)

for col, importance in critical_columns.items():
    if col in Job_Posting_df.columns:
        non_null = Job_Posting_df[col].notnull().sum()
        pct = (non_null / len(Job_Posting_df)) * 100
        
        if pct > 90:
            status = "✅ Excellent"
        elif pct > 70:
            status = "⚠️  Acceptable"
        elif pct > 50:
            status = "🔶 Concerning"
        else:
            status = "❌ Critical Issue"
        
        print(f"{col:23} | {non_null:8,d} | {pct:5.1f}% | {status}")
        print(f"                      {importance}")
    else:
        print(f"{col:23} | {'NOT FOUND':^8} | {'N/A':^5} | ❌ Missing Column")


CRITICAL COLUMNS ASSESSMENT

Column                  | Non-Null |   %   | Status
------------------------------------------------------------
Job Opening Title       |    9,919 | 100.0% | ✅ Excellent
                      Primary identifier - ESSENTIAL
Description             |    9,807 |  98.9% | ✅ Excellent
                      Contains skills/requirements - ESSENTIAL
Category                |    8,250 |  83.2% | ⚠️  Acceptable
                      Job classification - IMPORTANT
Location                |    9,508 |  95.9% | ✅ Excellent
                      Geographic info - IMPORTANT
Seniority               |    9,919 | 100.0% | ✅ Excellent
                      Experience level - IMPORTANT
Salary                  |      576 |   5.8% | ❌ Critical Issue
                      Compensation - DESIRABLE but limited
Contract Types          |    8,004 |  80.7% | ⚠️  Acceptable
                      Job type - DESIRABLE
Job Status              |    6,772 |  68.3% | 🔶 Concerning
         

- We can now see the most important columns which are desirable for our project and therefore we will go with this columns. Since most of our columns are text-based columns and they are categorical, we will have to develop key statistics which we will set for our categorical columns so that we can proceed with our data analysis.

In [12]:
## 2.4 Key Statistics for Numeric/Categorical Columns
print("\n" + "="*70)
print("CATEGORICAL COLUMNS ANALYSIS")
print("="*70)

categorical_cols = ['Category', 'Seniority', 'Job Status', 'Job Language', 'Contract Types']

for col in categorical_cols:
    if col in Job_Posting_df.columns and Job_Posting_df[col].notnull().sum() > 0:
        print(f"\n📊 {col}:")
        print("-"*40)
        
        # Count unique values
        unique_count = Job_Posting_df[col].nunique()
        non_null = Job_Posting_df[col].notnull().sum()
        
        print(f"Non-null values: {non_null:,}/{len(Job_Posting_df):,} ({(non_null/len(Job_Posting_df))*100:.1f}%)")
        print(f"Unique values: {unique_count}")
        
        # Show top values
        value_counts = Job_Posting_df[col].value_counts(dropna=False).head(10)
        print("\nTop 10 values:")
        for value, count in value_counts.items():
            pct = (count / len(Job_Posting_df)) * 100
            if pd.isna(value):
                print(f"  NaN: {count:5,d} ({pct:5.1f}%)")
            else:
                # Truncate long values
                display_value = str(value)[:50] + "..." if len(str(value)) > 50 else str(value)
                print(f"  {display_value:50}: {count:5,d} ({pct:5.1f}%)")
                


CATEGORICAL COLUMNS ANALYSIS

📊 Category:
----------------------------------------
Non-null values: 8,250/9,919 (83.2%)
Unique values: 509

Top 10 values:
  NaN: 1,669 ( 16.8%)
  engineering                                       :   986 (  9.9%)
  management                                        :   603 (  6.1%)
  internship                                        :   598 (  6.0%)
  manual_work                                       :   273 (  2.8%)
  software_development                              :   271 (  2.7%)
  engineering, quality_assurance                    :   185 (  1.9%)
  purchasing                                        :   182 (  1.8%)
  engineering, information_technology               :   177 (  1.8%)
  engineering, software_development                 :   171 (  1.7%)

📊 Seniority:
----------------------------------------
Non-null values: 9,919/9,919 (100.0%)
Unique values: 8

Top 10 values:
  non_manager                                       : 7,981 ( 80.5%)
  mana

- From this analysis, we can see that for the six categorical columns; i.e. , **Category**, **Seniority**, **Job Status**, **Job Language** and **Contract Types**, we have the various top values for each of these respective columns which shows us the Job Posting behaviour and nature at hand.

- Our dataset also happens to contain some columns which contains data in JSON format; i.e., ***Location Data*** and ***Salary Data***, hence the need to import the ***json*** library. Let's do a preview of the JSON columns.

In [13]:
# 2.5 JSON Columns Preview

print("\n" + "="*70)
print("JSON COLUMNS ANALYSIS")
print("="*70)

json_columns = ['Location Data', 'Salary Data']

for json_col in json_columns:
    if json_col in Job_Posting_df.columns:
        print(f"\n🔍 {json_col}:")
        print("-"*40)
        
        non_null_count = Job_Posting_df[json_col].notnull().sum()
        print(f"Non-null values: {non_null_count:,}/{len(Job_Posting_df):,} ({(non_null_count/len(Job_Posting_df))*100:.1f}%)")
        
        # Sample and parse JSON
        samples = Job_Posting_df[json_col].dropna().head(3)
        if len(samples) > 0:
            print("\nSample JSON structures:")
            for i, sample in enumerate(samples, 1):
                try:
                    if isinstance(sample, str) and sample.strip():
                        parsed = json.loads(sample)
                        print(f"\nSample {i}:")
                        if isinstance(parsed, list):
                            print(f"  Type: List with {len(parsed)} items")
                            if parsed and isinstance(parsed[0], dict):
                                print(f"  Keys in first item: {list(parsed[0].keys())}")
                        elif isinstance(parsed, dict):
                            print(f"  Type: Dictionary")
                            print(f"  Keys: {list(parsed.keys())}")
                            # Show first few key-value pairs
                            for key, value in list(parsed.items())[:3]:
                                print(f"    {key}: {str(value)[:50]}{'...' if len(str(value)) > 50 else ''}")
                    else:
                        print(f"Sample {i}: Empty or non-string value")
                except json.JSONDecodeError as e:
                    print(f"Sample {i}: Invalid JSON - {str(e)[:50]}")
                except Exception as e:
                    print(f"Sample {i}: Error - {type(e).__name__}: {str(e)[:50]}")


JSON COLUMNS ANALYSIS

🔍 Location Data:
----------------------------------------
Non-null values: 9,919/9,919 (100.0%)

Sample JSON structures:

Sample 1:
  Type: List with 1 items
  Keys in first item: ['city', 'state', 'zip_code', 'country', 'region', 'continent', 'fuzzy_match']

Sample 2:
  Type: List with 1 items
  Keys in first item: ['city', 'state', 'zip_code', 'country', 'region', 'continent', 'fuzzy_match']

Sample 3:
  Type: List with 1 items
  Keys in first item: ['city', 'state', 'zip_code', 'country', 'region', 'continent', 'fuzzy_match']

🔍 Salary Data:
----------------------------------------
Non-null values: 9,919/9,919 (100.0%)

Sample JSON structures:

Sample 1:
  Type: Dictionary
  Keys: ['salary_low', 'salary_high', 'salary_currency', 'salary_low_usd', 'salary_high_usd', 'salary_time_unit']
    salary_low: None
    salary_high: None
    salary_currency: None

Sample 2:
  Type: Dictionary
  Keys: ['salary_low', 'salary_high', 'salary_currency', 'salary_low_usd', 'sa

- The piece of code above was to identify the JSON columns so that we identify the various values and their categorical importance to the project and also identify the need to parse the columns.

- Now lets check through the text columns and the date columns;

In [14]:
# 2.6 Text Columns Preview

print("\n" + "="*70)
print("TEXT COLUMNS PREVIEW")
print("="*70)

text_columns = ['Job Opening Title', 'Description']

for col in text_columns:
    if col in Job_Posting_df.columns:
        print(f"\n📝 {col}:")
        print("-"*40)
        
        non_null = Job_Posting_df[col].notnull().sum()
        print(f"Non-null: {non_null:,}/{len(Job_Posting_df):,} ({(non_null/len(Job_Posting_df))*100:.1f}%)")
        
        # Show character statistics
        if non_null > 0:
            text_lengths = Job_Posting_df[col].dropna().apply(len)
            print(f"Average length: {text_lengths.mean():.0f} characters")
            print(f"Min length: {text_lengths.min():.0f} characters")
            print(f"Max length: {text_lengths.max():.0f} characters")
            
            print("\nSample entries:")
            samples = Job_Posting_df[col].dropna().head(3)
            for i, sample in enumerate(samples, 1):
                # Clean and truncate for display
                clean_sample = str(sample).replace('\n', ' ').replace('\r', ' ')
                if len(clean_sample) > 150:
                    display_text = clean_sample[:150] + "..."
                else:
                    display_text = clean_sample
                print(f"\n{i}. {display_text}")


TEXT COLUMNS PREVIEW

📝 Job Opening Title:
----------------------------------------
Non-null: 9,919/9,919 (100.0%)
Average length: 37 characters
Min length: 3 characters
Max length: 117 characters

Sample entries:

1. IN_RBAI_Assistant Manager_Dispensing Process Engineer_IN

2. Professional Internship: Hardware Development (M/F/Div.)

3. Process Expert BMS Production

📝 Description:
----------------------------------------
Non-null: 9,807/9,919 (98.9%)
Average length: 3401 characters
Min length: 165 characters
Max length: 8162 characters

Sample entries:

1. **IN\_RBAI\_Assistant Manager\_Dispensing Process Engineer\_IN**      * Full-time  * Legal Entity: Bosch Automotive Electronics India Private Ltd.    ...

2. **Professional Internship: Hardware Development (M/F/Div.)**      * Full-time  * Legal Entity: Home Comfort      **Company Description**    The Bosch ...

3. ZF is a global technology company supplying systems for passenger cars, commercial vehicles and industrial technology,

In [15]:
# 2.7 Date Columns Analysis

print("\n" + "="*70)
print("DATE COLUMNS ANALYSIS")
print("="*70)

date_columns = ['First Seen At', 'Last Seen At', 'Job Last Processed At']

for col in date_columns:
    if col in Job_Posting_df.columns:
        print(f"\n📅 {col}:")
        print("-"*40)
        
        # Check if already datetime
        if Job_Posting_df[col].dtype == 'object':
            # Try to convert
            try:
                temp_dates = pd.to_datetime(Job_Posting_df[col], errors='coerce')
                valid_dates = temp_dates.notnull().sum()
                print(f"Format appears to be: ISO 8601 (e.g., 2024-05-29T19:59:45Z)")
                print(f"Valid dates: {valid_dates:,}/{len(Job_Posting_df):,} ({(valid_dates/len(Job_Posting_df))*100:.1f}%)")
                
                if valid_dates > 0:
                    print(f"Date range: {temp_dates.min()} to {temp_dates.max()}")
                    duration_days = (temp_dates.max() - temp_dates.min()).days
                    print(f"Time span: {duration_days} days")
            except Exception as e:
                print(f"Conversion error: {str(e)[:50]}")
        else:
            print(f"Already datetime type")
            print(f"Date range: {Job_Posting_df[col].min()} to {Job_Posting_df[col].max()}")


DATE COLUMNS ANALYSIS

📅 First Seen At:
----------------------------------------
Format appears to be: ISO 8601 (e.g., 2024-05-29T19:59:45Z)
Valid dates: 9,919/9,919 (100.0%)
Date range: 2024-03-04 15:41:37+00:00 to 2024-09-04 07:03:16+00:00
Time span: 183 days

📅 Last Seen At:
----------------------------------------
Format appears to be: ISO 8601 (e.g., 2024-05-29T19:59:45Z)
Valid dates: 9,919/9,919 (100.0%)
Date range: 2024-03-06 16:31:21+00:00 to 2024-09-04 09:43:42+00:00
Time span: 181 days

📅 Job Last Processed At:
----------------------------------------
Format appears to be: ISO 8601 (e.g., 2024-05-29T19:59:45Z)
Valid dates: 9,919/9,919 (100.0%)
Date range: 2024-02-22 16:38:29+00:00 to 2024-09-04 09:43:42+00:00
Time span: 194 days


In [ ]:
- The code above show that the date and time columns for our dataset are good to go so we can now do a complete summary of the data quality of our dataset.

In [16]:
# ## 2.8 Data Quality Issues Summary

print("\n" + "="*70)
print("DATA QUALITY ISSUES SUMMARY")
print("="*70)

issues = []

# Issue 1: Completely missing column
if 'Ticker' in Job_Posting_df.columns and Job_Posting_df['Ticker'].isnull().all():
    issues.append(("❌ Ticker column", "100% missing - consider dropping"))

# Issue 2: Very high missing values
high_missing_cols = missing_df[missing_df['Null %'] > 50]['Column'].tolist()
for col in high_missing_cols:
    if col != 'Ticker' and col != 'Salary':  # Salary we already know about
        pct = missing_df[missing_df['Column'] == col]['Null %'].iloc[0]
        issues.append((f"⚠️  {col}", f"{pct:.1f}% missing"))

# Issue 3: Critical columns with significant missing data
for col in ['Category', 'Job Status']:
    if col in Job_Posting_df.columns:
        null_pct = (Job_Posting_df[col].isnull().sum() / len(Job_Posting_df)) * 100
        if null_pct > 20:
            issues.append((f"🔶 {col}", f"{null_pct:.1f}% missing - may affect analysis"))

# Issue 4: JSON parsing complexity
issues.append(("🔍 Location Data", "Requires JSON parsing for detailed location info"))
issues.append(("🔍 Salary Data", "Requires JSON parsing for structured salary info"))

print("\nIdentified Issues:")
print("-"*50)
if issues:
    for i, (issue, detail) in enumerate(issues, 1):
        print(f"{i:2}. {issue:25} - {detail}")
else:
    print("No major issues identified")


DATA QUALITY ISSUES SUMMARY

Identified Issues:
--------------------------------------------------
 1. ❌ Ticker column           - 100% missing - consider dropping
 2. 🔶 Job Status              - 31.7% missing - may affect analysis
 3. 🔍 Location Data           - Requires JSON parsing for detailed location info
 4. 🔍 Salary Data             - Requires JSON parsing for structured salary info


In [17]:
# 2.9 Recommendations for Next Steps

print("\n" + "="*70)
print("RECOMMENDED NEXT STEPS")
print("="*70)

next_steps = [
    ("1", "Drop completely empty columns", "Ticker column (0 non-null values)"),
    ("2", "Parse JSON columns", "Extract city, state, country from Location Data; salary details from Salary Data"),
    ("3", "Convert date columns", "Convert First Seen At, Last Seen At to datetime format"),
    ("4", "Handle missing Category data", "Consider imputation or separate 'unknown' category"),
    ("5", "Analyze text columns", "Extract skills from Description using NLP"),
    ("6", "Clean categorical columns", "Standardize values in Category, Seniority, Contract Types"),
    ("7", "Calculate posting duration", "Create new feature: Last Seen At - First Seen At"),
    ("8", "Explore Salary Data", "Extract and analyze available salary information"),
]

print("\n🛠️  Data Cleaning Priority:")
print("-"*50)
for num, step, details in next_steps:
    print(f"{num}. {step}")
    print(f"   → {details}")
    print()

print("\n" + "="*70)
print("PROJECT STATUS UPDATE")
print("="*70)
print(f"✅ Dataset loaded successfully: {Job_Posting_df.shape[0]:,} job postings")
print(f"✅ Critical columns identified and assessed")
print(f"✅ Data quality issues documented")
print(f"✅ Next steps outlined for cleaning and preparation")
print(f"\n📋 Ready for Step 3: Data Cleaning and Preparation")


RECOMMENDED NEXT STEPS

🛠️  Data Cleaning Priority:
--------------------------------------------------
1. Drop completely empty columns
   → Ticker column (0 non-null values)

2. Parse JSON columns
   → Extract city, state, country from Location Data; salary details from Salary Data

3. Convert date columns
   → Convert First Seen At, Last Seen At to datetime format

4. Handle missing Category data
   → Consider imputation or separate 'unknown' category

5. Analyze text columns
   → Extract skills from Description using NLP

6. Clean categorical columns
   → Standardize values in Category, Seniority, Contract Types

7. Calculate posting duration
   → Create new feature: Last Seen At - First Seen At

8. Explore Salary Data
   → Extract and analyze available salary information


PROJECT STATUS UPDATE
✅ Dataset loaded successfully: 9,919 job postings
✅ Critical columns identified and assessed
✅ Data quality issues documented
✅ Next steps outlined for cleaning and preparation

📋 Ready for

 3. Data Cleaning and Preparation

From our observaions, we noted that there were issues we needed to tackle so as to get the data ready for modelling. We decided to tackle the issues in this order;
- Drop completely empty columns

- Parse JSON columns (Location and Salary Data)

- Handle missing values

- Convert date columns

- Clean categorical/text data

- Create new features for the model

3.1 Initial Setup and Column removal

In [19]:
 #Make a copy for cleaning
Job_Posting_clean = Job_Posting_df.copy()
print("Initial shape:", Job_Posting_clean.shape)

# %%
print("\n" + "="*70)
print("3.1 DROP COMPLETELY EMPTY COLUMNS")
print("="*70)

# Drop Ticker column (100% missing)
if 'Ticker' in Job_Posting_clean.columns:
    Job_Posting_clean = Job_Posting_clean.drop(columns=['Ticker'])
    print("✅ Dropped 'Ticker' column (100% missing)")

print(f"New shape: {Job_Posting_clean.shape}")
print(f"Columns remaining: {len(Job_Posting_clean.columns)}")

Initial shape: (9919, 21)

3.1 DROP COMPLETELY EMPTY COLUMNS
✅ Dropped 'Ticker' column (100% missing)
New shape: (9919, 20)
Columns remaining: 20


 3.2 Parsing JSON columns

In [20]:
print("\n" + "="*70)
print("3.2 PARSE LOCATION DATA COLUMN")
print("="*70)

def parse_location_data(json_str):
    """Parse Location Data JSON and extract key fields"""
    try:
        if pd.isna(json_str) or json_str == '':
            return None, None, None, None, None
        
        data = json.loads(json_str)
        if isinstance(data, list) and len(data) > 0:
            location = data[0]
            return (
                location.get('city'),
                location.get('state'),
                location.get('country'),
                location.get('region'),
                location.get('continent')
            )
    except (json.JSONDecodeError, TypeError, KeyError) as e:
        pass
    return None, None, None, None, None

# Apply parsing
location_parsed = Job_Posting_clean['Location Data'].apply(parse_location_data)
Job_Posting_clean[['city', 'state', 'country', 'region', 'continent']] = pd.DataFrame(
    location_parsed.tolist(), index=Job_Posting_clean.index
)

print("✅ Extracted location fields from Location Data:")
print(f"   - city: {Job_Posting_clean['city'].notnull().sum():,} non-null")
print(f"   - state: {Job_Posting_clean['state'].notnull().sum():,} non-null")
print(f"   - country: {Job_Posting_clean['country'].notnull().sum():,} non-null")
print(f"   - region: {Job_Posting_clean['region'].notnull().sum():,} non-null")
print(f"   - continent: {Job_Posting_clean['continent'].notnull().sum():,} non-null")

# Show sample
print("\n📊 Sample extracted location data:")
sample_idx = Job_Posting_clean[Job_Posting_clean['country'].notnull()].index[0]
print(f"Original Location: {Job_Posting_clean.loc[sample_idx, 'Location']}")
print(f"Parsed - City: {Job_Posting_clean.loc[sample_idx, 'city']}")
print(f"Parsed - State: {Job_Posting_clean.loc[sample_idx, 'state']}")
print(f"Parsed - Country: {Job_Posting_clean.loc[sample_idx, 'country']}")


3.2 PARSE LOCATION DATA COLUMN
✅ Extracted location fields from Location Data:
   - city: 6,281 non-null
   - state: 2,450 non-null
   - country: 9,449 non-null
   - region: 21 non-null
   - continent: 33 non-null

📊 Sample extracted location data:
Original Location: Indiana, United States
Parsed - City: None
Parsed - State: Indiana
Parsed - Country: United States


In [21]:
print("\n" + "="*70)
print("3.3 PARSE SALARY DATA COLUMN")
print("="*70)

def parse_salary_data(json_str):
    """Parse Salary Data JSON and extract key fields"""
    try:
        if pd.isna(json_str) or json_str == '':
            return None, None, None, None, None, None
        
        data = json.loads(json_str)
        return (
            data.get('salary_low'),
            data.get('salary_high'),
            data.get('salary_currency'),
            data.get('salary_low_usd'),
            data.get('salary_high_usd'),
            data.get('salary_time_unit')
        )
    except (json.JSONDecodeError, TypeError, KeyError) as e:
        pass
    return None, None, None, None, None, None

# Apply parsing
salary_parsed = Job_Posting_clean['Salary Data'].apply(parse_salary_data)
Job_Posting_clean[['salary_low', 'salary_high', 'salary_currency', 
          'salary_low_usd', 'salary_high_usd', 'salary_time_unit']] = pd.DataFrame(
    salary_parsed.tolist(), index=Job_Posting_clean.index
)

print("✅ Extracted salary fields from Salary Data:")
salary_fields = ['salary_low', 'salary_high', 'salary_currency', 
                 'salary_low_usd', 'salary_high_usd', 'salary_time_unit']
for field in salary_fields:
    non_null = Job_Posting_clean[field].notnull().sum()
    print(f"   - {field:20}: {non_null:6,} non-null ({non_null/len(Job_Posting_clean)*100:.1f}%)")

# Check if we have any actual salary data
has_salary_data = Job_Posting_clean['salary_low'].notnull().sum() > 0
print(f"\n🔍 Salary data availability: {'✅ Yes' if has_salary_data else '❌ No actual salary values found'}")


3.3 PARSE SALARY DATA COLUMN
✅ Extracted salary fields from Salary Data:
   - salary_low          :    434 non-null (4.4%)
   - salary_high         :    434 non-null (4.4%)
   - salary_currency     :    434 non-null (4.4%)
   - salary_low_usd      :    434 non-null (4.4%)
   - salary_high_usd     :    434 non-null (4.4%)
   - salary_time_unit    :    434 non-null (4.4%)

🔍 Salary data availability: ✅ Yes


 3.4 Converting Date Columns

In [22]:
print("\n" + "="*70)
print("3.4 CONVERT DATE COLUMNS")
print("="*70)

date_columns = ['First Seen At', 'Last Seen At', 'Job Last Processed At']

for col in date_columns:
    if col in Job_Posting_clean.columns:
        Job_Posting_clean[col] = pd.to_datetime(Job_Posting_clean[col], errors='coerce', utc=True)
        valid_dates = Job_Posting_clean[col].notnull().sum()
        print(f"✅ Converted {col:25}: {valid_dates:,} valid dates")
        
        # Show date range
        if valid_dates > 0:
            min_date = Job_Posting_clean[col].min()
            max_date = Job_Posting_clean[col].max()
            print(f"   Range: {min_date.strftime('%Y-%m-%d')} to {max_date.strftime('%Y-%m-%d')}")

# Create new feature: Job posting duration (in days)
if 'First Seen At' in Job_Posting_clean.columns and 'Last Seen At' in Job_Posting_clean.columns:
   Job_Posting_clean['posting_duration_days'] = (Job_Posting_clean['Last Seen At'] - Job_Posting_clean['First Seen At']).dt.days
   print(f"\n✅ Created new feature: posting_duration_days")
   print(f"   Average duration: {Job_Posting_clean['posting_duration_days'].mean():.1f} days")
   print(f"   Min duration: {Job_Posting_clean['posting_duration_days'].min():.1f} days")
   print(f"   Max duration: {Job_Posting_clean['posting_duration_days'].max():.1f} days")


3.4 CONVERT DATE COLUMNS
✅ Converted First Seen At            : 9,919 valid dates
   Range: 2024-03-04 to 2024-09-04
✅ Converted Last Seen At             : 9,919 valid dates
   Range: 2024-03-06 to 2024-09-04
✅ Converted Job Last Processed At    : 9,919 valid dates
   Range: 2024-02-22 to 2024-09-04

✅ Created new feature: posting_duration_days
   Average duration: 39.3 days
   Min duration: 0.0 days
   Max duration: 182.0 days


3.5 Handling Missing Values

In [23]:
print("\n" + "="*70)
print("3.5 HANDLE MISSING VALUES")
print("="*70)

# Track missing values before handling
missing_before = Job_Posting_clean.isnull().sum().sort_values(ascending=False)
print("Missing values before handling (top 10):")
print(missing_before.head(10))


3.5 HANDLE MISSING VALUES
Missing values before handling (top 10):
region              9898
continent           9886
salary_time_unit    9485
salary_high_usd     9485
salary_low_usd      9485
salary_currency     9485
salary_high         9485
salary_low          9485
Salary              9343
state               7469
dtype: int64


In [24]:
print("\n" + "="*70)
print("MISSING VALUE HANDLING STRATEGY")
print("="*70)

# Strategy for each column
missing_strategies = {
    'Category': "Fill with 'unknown' category",
    'Job Status': "Fill with 'unknown' status",
    'Keywords': "Fill with empty string",
    'Contract Types': "Fill with 'not_specified'",
    'Location': "Keep as is (95.9% complete), fill with 'Unknown'",
    'Description': "Drop rows (only 112 missing)",
    'city': "Keep parsed values (some will be null)",
    'state': "Keep parsed values",
    'country': "Keep parsed values",
    'salary_low': "Keep as is (salary data is sparse)"
}

print("\nHandling strategy for key columns:")
print("-"*50)
for col, strategy in missing_strategies.items():
    if col in Job_Posting_clean.columns:
        missing = Job_Posting_clean[col].isnull().sum()
        pct = (missing / len(Job_Posting_clean)) * 100
        print(f"{col:20} | {missing:5,} missing ({pct:5.1f}%) → {strategy}")


MISSING VALUE HANDLING STRATEGY

Handling strategy for key columns:
--------------------------------------------------
Category             | 1,669 missing ( 16.8%) → Fill with 'unknown' category
Job Status           | 3,147 missing ( 31.7%) → Fill with 'unknown' status
Keywords             | 2,273 missing ( 22.9%) → Fill with empty string
Contract Types       | 1,915 missing ( 19.3%) → Fill with 'not_specified'
Location             |   411 missing (  4.1%) → Keep as is (95.9% complete), fill with 'Unknown'
Description          |   112 missing (  1.1%) → Drop rows (only 112 missing)
city                 | 3,638 missing ( 36.7%) → Keep parsed values (some will be null)
state                | 7,469 missing ( 75.3%) → Keep parsed values
country              |   470 missing (  4.7%) → Keep parsed values
salary_low           | 9,485 missing ( 95.6%) → Keep as is (salary data is sparse)


In [25]:
# Apply missing value handling
print("\n" + "="*70)
print("APPLYING MISSING VALUE HANDLING")
print("="*70)

# Fill categorical columns
Job_Posting_clean['Category'] = Job_Posting_clean['Category'].fillna('unknown')
Job_Posting_clean['Job Status'] = Job_Posting_clean['Job Status'].fillna('unknown')
Job_Posting_clean['Keywords'] = Job_Posting_clean['Keywords'].fillna('')
Job_Posting_clean['Contract Types'] = Job_Posting_clean['Contract Types'].fillna('not_specified')
Job_Posting_clean['Location'] = Job_Posting_clean['Location'].fillna('Unknown')

# For Description, we have very few missing, so we can drop
rows_before = len(Job_Posting_clean)
Job_Posting_clean = Job_Posting_clean.dropna(subset=['Description'])
rows_after = len(Job_Posting_clean)
print(f"✅ Dropped {rows_before - rows_after} rows with missing Description")

print("\nMissing values after handling (top 10):")
missing_after = Job_Posting_clean.isnull().sum().sort_values(ascending=False)
print(missing_after.head(10))


APPLYING MISSING VALUE HANDLING
✅ Dropped 112 rows with missing Description

Missing values after handling (top 10):
region              9786
continent           9774
salary_time_unit    9373
salary_high_usd     9373
salary_low_usd      9373
salary_currency     9373
salary_high         9373
salary_low          9373
Salary              9231
state               7381
dtype: int64


 3.6 Standardize Categorical Columns

In [26]:
print("\n" + "="*70)
print("3.6 CLEAN CATEGORICAL COLUMNS")
print("="*70)

# Clean Category column - split multiple categories
print("🔧 Cleaning 'Category' column...")
Job_Posting_clean['Category_list'] = Job_Posting_clean['Category'].apply(
    lambda x: [cat.strip() for cat in str(x).split(',')] if pd.notnull(x) else []
)

# Create indicator for single vs multiple categories
Job_Posting_clean['has_multiple_categories'] = Job_Posting_clean['Category_list'].apply(lambda x: len(x) > 1)

print(f"✅ Created Category_list and has_multiple_categories features")
print(f"   Jobs with multiple categories: {Job_Posting_clean['has_multiple_categories'].sum():,} ({Job_Posting_clean['has_multiple_categories'].mean()*100:.1f}%)")


3.6 CLEAN CATEGORICAL COLUMNS
🔧 Cleaning 'Category' column...
✅ Created Category_list and has_multiple_categories features
   Jobs with multiple categories: 3,994 (40.7%)


In [27]:
# Clean Seniority column
print("\n🔧 Cleaning 'Seniority' column...")
seniority_mapping = {
    'non_manager': 'individual_contributor',
    'manager': 'manager',
    'head': 'director_level',
    'director': 'director_level',
    'c_level': 'executive',
    'vice_president': 'executive',
    'partner': 'executive',
    'president': 'executive'
}

Job_Posting_clean['Seniority_clean'] = Job_Posting_clean['Seniority'].map(seniority_mapping)
Job_Posting_clean['Seniority_clean'] = Job_Posting_clean['Seniority_clean'].fillna('other')

print("✅ Standardized Seniority levels:")
print(Job_Posting_clean['Seniority_clean'].value_counts())


🔧 Cleaning 'Seniority' column...
✅ Standardized Seniority levels:
Seniority_clean
individual_contributor    7889
manager                   1791
director_level             107
executive                   20
Name: count, dtype: int64


In [28]:
# Clean Contract Types
print("\n🔧 Cleaning 'Contract Types' column...")

# Extract primary contract type (first one if multiple)
def extract_primary_contract(contract_str):
    if pd.isna(contract_str) or contract_str == 'not_specified':
        return 'not_specified'
    
    # Split by comma and take first
    contracts = str(contract_str).split(',')
    primary = contracts[0].strip().lower()
    
    # Map to standard terms
    contract_mapping = {
        'full time': 'full_time',
        'part time': 'part_time',
        'intern': 'internship',
        'vollzeit': 'full_time',  # German
        'tempo integral': 'full_time',  # Portuguese
        'm/f': 'full_time',  # Probably means full-time
        'm/w': 'full_time',  # Probably means full-time
        'hybrid': 'hybrid'
    }
    
    return contract_mapping.get(primary, primary)

Job_Posting_clean['Contract_Type_primary'] = Job_Posting_clean['Contract Types'].apply(extract_primary_contract)

print("✅ Primary contract types:")
print(Job_Posting_clean['Contract_Type_primary'].value_counts().head(10))


🔧 Cleaning 'Contract Types' column...
✅ Primary contract types:
Contract_Type_primary
full_time        5348
not_specified    1902
internship        741
hybrid            434
part_time         188
long term         179
all levels        176
contract          174
remote            170
permanent          83
Name: count, dtype: int64


 3.7 Cleaning Text Columns

In [29]:
print("\n" + "="*70)
print("3.7 CLEAN TEXT COLUMNS")
print("="*70)

# Clean Job Opening Title
print("🔧 Cleaning 'Job Opening Title'...")

# Remove extra whitespace and standardize case
Job_Posting_clean['Title_clean'] = Job_Posting_clean['Job Opening Title'].str.strip().str.lower()

# Extract potential indicators from title
Job_Posting_clean['title_has_senior'] = Job_Posting_clean['Title_clean'].str.contains('senior', case=False)
Job_Posting_clean['title_has_junior'] = Job_Posting_clean['Title_clean'].str.contains('junior', case=False)
Job_Posting_clean['title_has_manager'] = Job_Posting_clean['Title_clean'].str.contains('manager', case=False)
Job_Posting_clean['title_has_engineer'] = Job_Posting_clean['Title_clean'].str.contains('engineer', case=False)
Job_Posting_clean['title_has_developer'] = Job_Posting_clean['Title_clean'].str.contains('developer', case=False)
Job_Posting_clean['title_has_analyst'] = Job_Posting_clean['Title_clean'].str.contains('analyst', case=False)

print("✅ Title indicators extracted:")
indicators = ['title_has_senior', 'title_has_junior', 'title_has_manager', 
              'title_has_engineer', 'title_has_developer', 'title_has_analyst']
for indicator in indicators:
    count = Job_Posting_clean[indicator].sum()
    print(f"   - {indicator:20}: {count:6,} ({count/len(Job_Posting_clean)*100:.1f}%)")


3.7 CLEAN TEXT COLUMNS
🔧 Cleaning 'Job Opening Title'...
✅ Title indicators extracted:
   - title_has_senior    :    630 (6.4%)
   - title_has_junior    :     78 (0.8%)
   - title_has_manager   :  1,044 (10.6%)
   - title_has_engineer  :  1,902 (19.4%)
   - title_has_developer :    351 (3.6%)
   - title_has_analyst   :    361 (3.7%)


In [30]:
# Initial Description cleaning
print("\n🔧 Initial cleaning of 'Description'...")

# Store original length
Job_Posting_clean['Description_length'] = Job_Posting_clean['Description'].str.len()

# Basic cleaning: remove extra whitespace
Job_Posting_clean['Description_clean'] = Job_Posting_clean['Description'].str.replace(r'\s+', ' ', regex=True).str.strip()

print(f"✅ Description length statistics:")
print(f"   Average: {Job_Posting_clean['Description_length'].mean():.0f} characters")
print(f"   Min: {Job_Posting_clean['Description_length'].min():.0f} characters")
print(f"   Max: {Job_Posting_clean['Description_length'].max():.0f} characters")


🔧 Initial cleaning of 'Description'...
✅ Description length statistics:
   Average: 3401 characters
   Min: 165 characters
   Max: 8162 characters
